packages/devtools_app/lib/src/memory/flutter/memory_heap_tree_view.dart

```diff
@@ -0,0 +1,848 @@
+
+  Widget snapshotDisplay;
+
+  @override
+  void initState() {
+    super.initState();
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+
+    controller = Provider.of<MemoryController>(context);
+
+    cancel();
+
+    addAutoDisposeListener(controller.selectedSnapshotNotifier, () {
+      setState(() {
+        controller.computeAllLibraries(true, true);
+      });
+    });
+
+    addAutoDisposeListener(controller.filterNotifier, () {
+      setState(() {
+        controller.computeAllLibraries(true, true);
+      });
+    });
+
+    addAutoDisposeListener(controller.leafSelectedNotifier, () {
+      setState(() {});
+    });
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    // TODO(terry): Make HeatMapSizeAnalyzer a stateful widget.
+    heapGraph = controller.theGraph;
+    snapshotDisplay = controller.snapshotByLibraryData == null
+        ? const Text('')
+        : controller.showHeatMap
+            ? HeatMapSizeAnalyzer()
+            : MemorySnapshotTable();
+
+    return Column(
+      children: [
+        Row(
+          mainAxisAlignment: MainAxisAlignment.spaceBetween,
+          children: [
+            _buildSnapshotControls(),
+            _buildSearchFilterControls(),
+          ],
+        ),
+        Expanded(
+          child: buildSnapshotTables(snapshotDisplay),
+        ),
+      ],
+    );
+  }
+
+  Widget buildSnapshotTables(Widget snapshotDisplay) {
+    return Container(
+      padding: const EdgeInsets.symmetric(horizontal: 6),
+      child: Row(
+        mainAxisSize: MainAxisSize.min,
+        mainAxisAlignment: MainAxisAlignment.end,
+        children: [
+          Expanded(child: snapshotDisplay),
+          const SizedBox(width: 16.0),
+          controller.isLeafSelected
+              ? Expanded(child: SnapshotInstanceViewTable())
+              : const Text(''),
+        ],
+      ),
+    );
+  }
+
+  @visibleForTesting
+  static const groupByMenuButtonKey = Key('Group By Menu Button');
+  @visibleForTesting
+  static const groupByMenuItem = Key('Filter Group By Menu Item');
+  @visibleForTesting
+  static const groupByKey = Key('Filter Group By');
+
+  Widget _groupByDropdown() {
+    final _groupByTypes = [
+      MemoryController.groupByLibrary,
+      MemoryController.groupByClass,
+      MemoryController.groupByInstance,
+    ].map<DropdownMenuItem<String>>(
+      (
+        String value,
+      ) {
+        return DropdownMenuItem<String>(
+          key: groupByMenuItem,
+          value: value,
+          child: Text(
+            'Group by $value',
+            key: groupByKey,
+          ),
+        );
+      },
+    ).toList();
+
+    return DropdownButton<String>(
+      key: groupByMenuButtonKey,
+      value: controller.groupingBy.value,
+      iconSize: 20,
+      style: const TextStyle(fontWeight: FontWeight.w100),
+      onChanged: (String newValue) {
+        setState(
```

Comment:
in general, calls to a controller should be outside a setState block as the controller can be modified without calling setState on a particular view using it.